plugins/notifications-backend/src/service/auth.ts

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2024 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { AuthenticationError, NotAllowedError } from '@backstage/errors';
+import { getBearerTokenFromAuthorizationHeader } from '@backstage/plugin-auth-node';
+import {
+  AuthorizeResult,
+  BasicPermission,
+  PermissionEvaluator,
+} from '@backstage/plugin-permission-common';
+
+import express from 'express';
+
+import { DefaultServiceUser } from './constants';
+import { RouterOptions } from './types';
+
+export type GetLoggedInUserOptions = Pick<
+  RouterOptions,
+  'identity' | 'tokenManager' | 'externalCallerSecret'
+>;
+export type CheckUserPermission = GetLoggedInUserOptions &
+  Pick<RouterOptions, 'permissions'>;
+
+/*
+ * User's entity must be present in the catalog.
+ */
+export const getLoggedInUser = async (
+  request: express.Request,
+  { identity, tokenManager, externalCallerSecret }: GetLoggedInUserOptions,
+): Promise<string> => {
+  const identityResponse = await identity.getIdentity({ request });
+
+  // To properly set identity, see packages/backend/src/plugins/auth.ts or https://backstage.io/docs/auth/identity-resolver
+  if (identityResponse) {
+    // The auth token contains user's identity, most probably originated in the FE
+    // Example: user:default/guest
+    let author = identityResponse?.identity.userEntityRef;
+    if (author) {
+      if (author.startsWith('user:')) {
+        author = author.slice('user:'.length);
+      }
+    } else {
+      throw new AuthenticationError(
+        'Missing valid authentication data or the user is not in the Catalog.',
+      );
+    }
+    return author;
+  }
+
+  const token = getBearerTokenFromAuthorizationHeader(
+    request.header('authorization'),
+  );
+  if (token) {
+    // backend service-to-service flow
+    await tokenManager.authenticate(token);
+  }
+
+  // External call - workaround
+  // Following shared-secret is a workaround till we make the creation of valid JWT tokens by external callers simple.
+  // In such case, the flow would be identical with the service-to-service.
+  // https://github.com/backstage/backstage/issues/18622
+  // https://github.com/backstage/backstage/issues/9374
+  if (externalCallerSecret) {
+    if (request.header('notifications-secret') === externalCallerSecret) {
```

Comment:
This is very custom functionality and probably should not be here. I think there is way to handle outside calls to backstage APIs some different way, @freben or @Rugvip any comments?

Reply:
It should be possible to create custom JWT tokens. So far I did not manage to make it working.

This is a workaround using shared secrets. We might reuse the shared secrets with the `auth` plugin but this seems to be too dangerous to me. I would prefer to use one or more specific ones.

Reply:
If we figure our a clear instruction, how a 3rd party can get/create/maintain/renew a valid JWT token, we can remove that.